### ISSUE TITLE
[arm32v7]  trusty-security does not exist for armhf?

### ISSUE THREAD

**Issue by ruffsl** [tianon/docker-brew-ubuntu-core#94]
Not sure if this is the place to report, but I think the `arm32v7` version of the official ubuntu image on docker hub has a source list that is trying to access trusty-security for an architecture not listed. 
http://security.ubuntu.com/ubuntu/dists/trusty-security/

Perhaps this is some residue from templating the imaging over multiple ubuntu architectures?
This came up during our builds for ROS on arm32v7, just as we apt update with no changes to base:
https://doi-janky.infosiftr.net/job/multiarch/job/arm32v7/job/ros/1/console

Using [cross-docker](https://github.com/justincormack/cross-docker) to reproduce the issue:
```
$ ./cross-docker run -it --rm arm32v7/ubuntu:trusty apt-update
Unable to find image 'arm32v7/ubuntu:trusty' locally
trusty: Pulling from arm32v7/ubuntu
3fe87bc44db5: Pull complete 
35850dbe54d8: Pull complete 
fb6a1cfbf532: Pull complete 
98d1baa4f111: Pull complete 
84e4a7e7c0cf: Pull complete 
Digest: sha256:51f90d9f8181335664f72cea7ddfd7df1fda6361d55001eee67d2f21cf955a03
Status: Downloaded newer image for arm32v7/ubuntu:trusty
docker: Error response from daemon: oci runtime error: container_linux.go:262: starting container process caused "exec: \"apt-update\": executable file not found in $PATH".
ruffsl@eve:~/git/justincormack/cross-docker$ ./cross-docker run -it --rm arm32v7/ubuntu:trusty bash -c "apt-update"
/bin/bash: apt-update: command not found
ruffsl@eve:~/git/justincormack/cross-docker$ ./cross-docker run -it --rm arm32v7/ubuntu:trusty apt-get update
Get:1 http://security.ubuntu.com trusty-security InRelease [65.9 kB]
Ign http://ports.ubuntu.com trusty InRelease               
Get:2 http://ports.ubuntu.com trusty-updates InRelease [65.9 kB]
Get:3 http://ports.ubuntu.com trusty-backports InRelease [65.9 kB]
Get:4 http://ports.ubuntu.com trusty Release.gpg [933 B]
Get:5 http://ports.ubuntu.com trusty Release [58.5 kB]          
Get:6 http://security.ubuntu.com trusty-security/universe Sources [73.6 kB]
Get:7 http://ports.ubuntu.com trusty-updates/universe Sources [237 kB]         
Err http://security.ubuntu.com trusty-security/main armhf Packages             
  404  Not Found [IP: 91.189.91.26 80]
Err http://security.ubuntu.com trusty-security/restricted armhf Packages       
  404  Not Found [IP: 91.189.91.26 80]
Err http://security.ubuntu.com trusty-security/universe armhf Packages         
  404  Not Found [IP: 91.189.91.26 80]
Get:8 http://ports.ubuntu.com trusty-updates/main armhf Packages [1060 kB] 
Err http://security.ubuntu.com trusty-security/multiverse armhf Packages       
  404  Not Found [IP: 91.189.91.26 80]
Get:9 http://ports.ubuntu.com trusty-updates/restricted armhf Packages [10.4 kB]
Get:10 http://ports.ubuntu.com trusty-updates/universe armhf Packages [531 kB] 
Get:11 http://ports.ubuntu.com trusty-updates/multiverse armhf Packages [11.2 kB]
Get:12 http://ports.ubuntu.com trusty-backports/main armhf Packages [14.8 kB]  
Get:13 http://ports.ubuntu.com trusty-backports/restricted armhf Packages [40 B]
Get:14 http://ports.ubuntu.com trusty-backports/universe armhf Packages [52.3 kB]
Get:15 http://ports.ubuntu.com trusty-backports/multiverse armhf Packages [871 B]
Get:16 http://ports.ubuntu.com trusty/universe Sources [7926 kB]               
Get:17 http://ports.ubuntu.com trusty/main armhf Packages [1672 kB]            
Get:18 http://ports.ubuntu.com trusty/restricted armhf Packages [40 B]         
Get:19 http://ports.ubuntu.com trusty/universe armhf Packages [7393 kB]        
Get:20 http://ports.ubuntu.com trusty/multiverse armhf Packages [142 kB]       
Fetched 19.4 MB in 31s (606 kB/s)                                              
W: Failed to fetch http://security.ubuntu.com/ubuntu/dists/trusty-security/main/binary-armhf/Packages  404  Not Found [IP: 91.189.91.26 80]

W: Failed to fetch http://security.ubuntu.com/ubuntu/dists/trusty-security/restricted/binary-armhf/Packages  404  Not Found [IP: 91.189.91.26 80]

W: Failed to fetch http://security.ubuntu.com/ubuntu/dists/trusty-security/universe/binary-armhf/Packages  404  Not Found [IP: 91.189.91.26 80]

W: Failed to fetch http://security.ubuntu.com/ubuntu/dists/trusty-security/multiverse/binary-armhf/Packages  404  Not Found [IP: 91.189.91.26 80]

E: Some index files failed to download. They have been ignored, or old ones used instead.
```

Inspecting the source.list:
```
$ ./cross-docker run -it --rm arm32v7/ubuntu:trusty cat /etc/apt/sources.list
# See http://help.ubuntu.com/community/UpgradeNotes for how to upgrade to
# newer versions of the distribution.
deb http://ports.ubuntu.com/ubuntu-ports/ trusty main restricted
# deb-src http://ports.ubuntu.com/ubuntu-ports/ trusty main restricted

## Major bug fix updates produced after the final release of the
## distribution.
deb http://ports.ubuntu.com/ubuntu-ports/ trusty-updates main restricted
# deb-src http://ports.ubuntu.com/ubuntu-ports/ trusty-updates main restricted

## N.B. software from this repository is ENTIRELY UNSUPPORTED by the Ubuntu
## team. Also, please note that software in universe WILL NOT receive any
## review or updates from the Ubuntu security team.
deb http://ports.ubuntu.com/ubuntu-ports/ trusty universe
deb-src http://ports.ubuntu.com/ubuntu-ports/ trusty universe
deb http://ports.ubuntu.com/ubuntu-ports/ trusty-updates universe
deb-src http://ports.ubuntu.com/ubuntu-ports/ trusty-updates universe

## N.B. software from this repository is ENTIRELY UNSUPPORTED by the Ubuntu
## team, and may not be under a free licence. Please satisfy yourself as to
## your rights to use the software. Also, please note that software in
## multiverse WILL NOT receive any review or updates from the Ubuntu
## security team.
deb http://ports.ubuntu.com/ubuntu-ports/ trusty multiverse
# deb-src http://ports.ubuntu.com/ubuntu-ports/ trusty multiverse
deb http://ports.ubuntu.com/ubuntu-ports/ trusty-updates multiverse
# deb-src http://ports.ubuntu.com/ubuntu-ports/ trusty-updates multiverse

## N.B. software from this repository may not have been tested as
## extensively as that contained in the main release, although it includes
## newer versions of some applications which may provide useful features.
## Also, please note that software in backports WILL NOT receive any review
## or updates from the Ubuntu security team.
deb http://ports.ubuntu.com/ubuntu-ports/ trusty-backports main restricted universe multiverse
# deb-src http://ports.ubuntu.com/ubuntu-ports/ trusty-backports main restricted universe multiverse

## Uncomment the following two lines to add software from Canonical's
## 'partner' repository.
## This software is not part of Ubuntu, but is offered by Canonical and the
## respective vendors as a service to Ubuntu users.
# deb http://archive.canonical.com/ubuntu trusty partner
# deb-src http://archive.canonical.com/ubuntu trusty partner

deb http://security.ubuntu.com/ubuntu trusty-security main restricted
# deb-src http://security.ubuntu.com/ubuntu trusty-security main restricted
deb http://security.ubuntu.com/ubuntu trusty-security universe
deb-src http://security.ubuntu.com/ubuntu trusty-security universe
deb http://security.ubuntu.com/ubuntu trusty-security multiverse
# deb-src http://security.ubuntu.com/ubuntu trusty-security multiverse
```

pinging @yosifkit @tianon @tfoote @mikaelarguedas

**Comment by tfoote** [comment]
The url it's trying only works for amd64 and i386
http://security.ubuntu.com/ubuntu/dists/trusty-security/main/

For arm64 armhf and other architectures it can be found at: 
http://security.ubuntu.com/ubuntu/dists/trusty-security/main/

It's one of the idiosyncrasies of the ubuntu multi-arch hosting scheme.

**Comment by ruffsl** [comment]
> For arm64 armhf and other architectures it can be found at:
http://security.ubuntu.com/ubuntu/dists/trusty-security/main/

Did you mean?:
http://ports.ubuntu.com/dists/trusty-security/

**Comment by mikaelarguedas** [comment]
Yeah as discussed yesterday all arm packages are hosted on ports.ubuntu.com/ubuntu-ports.

I didn't see in the code where the sources are modified but I would expect that somewhere
`http://archive.ubuntu.com/ubuntu/` is replaced by `http://ports.ubuntu.com/ubuntu-ports/`
To fix this issue that piece of code should also replace:
`http://security.ubuntu.com/ubuntu` with `http://ports.ubuntu.com/ubuntu-ports`

**Comment by tianon** [comment]
> Not sure if this is the place to report, ...

From https://hub.docker.com/_/ubuntu/:

> -	**Where to file issues**:  
> 	[the cloud-images bug tracker](https://bugs.launchpad.net/cloud-images) (include the `docker` tag)

Especially for this, since that `sources.list` is created by Canonical in the https://partner-images.canonical.com/core/ tarballs they generate; all we do here is adjust it to include universe by default:
https://github.com/tianon/docker-brew-ubuntu-core/blob/a2573609340194bf33712c1fa2dc4de8f9b70ca2/xenial/Dockerfile#L39-L40

**Comment by ruffsl** [comment]
Yea, I saw that but then second guess myself, as last I investigated the origins of the ubuntu, it seemed like this used to have a few more custom tweaks with a debootstrap phase.

https://github.com/tianon/docker-brew-ubuntu-core/blob/a2573609340194bf33712c1fa2dc4de8f9b70ca2/xenial/Dockerfile#L39-L40

Yep, nothing awry here then. I'm in a bit of a state of disbelief that this cloud image is shipped like this.

**Comment by ruffsl** [comment]
I think https://bugs.launchpad.net/cloud-images/+bug/1711735 has been resolved.
Just checked the official library though, but it doesn't seem to be from the latest rootfs available.

Should we raise a new launchpad issue asking if the "current" folder in Index of /core/trusty at https://partner-images.canonical.com/core/trusty/ be updated to redirect to 20171102 instead of 20170817, as it is presently, or is that an argument parameter you bump when building here?

**Comment by tianon** [comment]
@gaughen it looks like the most recent trusty build with @mwhudson's fix (20171102) failed on ppc64el so didn't generate `SHA*SUMS`, signatures, or update the `current` symlink -- does it just need to be triggered again or is something else wrong now? :pray: :heart:

**Comment by mwhudson** [comment]
On 7/11/2017 17:23, "Tianon Gravi" <notifications@github.com> wrote:

@gaughen <https://github.com/gaughen> it looks like the most recent trusty
build with @mwhudson <https://github.com/mwhudson>'s fix (20171102) failed
on ppc64el so didn't generate SHA*SUMS, signatures, or update the current
symlink -- does it just need to be triggered again or is something else
wrong now? 🙏

I think it just needs to be triggered again. The bug causing the failures
on arm64 and ppc64el has been fixed, or at least so people claim 😉
